database/service_db.go

```diff
@@ -52,59 +75,105 @@ func (d *LevelDBServiceDB) unmarshal(id string, value []byte) (*service.Service,
 func (d *LevelDBServiceDB) All() ([]*service.Service, error) {
 	var (
 		services []*service.Service
-		iter     = d.db.NewIterator(nil, nil)
+		iter     = d.db.NewIterator(util.BytesPrefix([]byte(idKeyPrefix)), nil)
 	)
-
 	for iter.Next() {
-		s, err := d.unmarshal(string(iter.Key()), iter.Value())
+		id := strings.TrimPrefix(string(iter.Key()), idKeyPrefix)
+		s, err := d.unmarshal(id, iter.Value())
 		if err != nil {
 			// NOTE: Ignore all decode errors (possibly due to a service
 			// structure change or database corruption)
 			if decodeErr, ok := err.(*DecodeError); ok {
 				logrus.WithField("service", decodeErr.ID).Warning(decodeErr.Error())
 				continue
 			}
+			iter.Release()
 			return nil, err
 		}
 		services = append(services, s)
 	}
 	iter.Release()
-	if err := iter.Error(); err != nil {
-		return nil, err
-	}
-
-	return services, nil
+	return services, iter.Error()
 }
 
 // Delete deletes service from database.
-func (d *LevelDBServiceDB) Delete(id string) error {
-	return d.db.Delete([]byte(id), nil)
+func (d *LevelDBServiceDB) Delete(idOrAlias string) error {
+	tx, err := d.db.OpenTransaction()
+	if err != nil {
+		return err
+	}
+
+	s, err := d.get(tx, idOrAlias)
+	if err != nil {
+		tx.Discard()
+		return err
+	}
+
+	batch := &leveldb.Batch{}
+	batch.Delete([]byte(idKeyPrefix + s.ID))
+	batch.Delete([]byte(aliasKeyPrefix + s.Alias))
+	if err := tx.Write(batch, nil); err != nil {
+		tx.Discard()
+		return err
+	}
+	if err := tx.Commit(); err != nil {
+		tx.Discard()
```

Comment:
@krhubert @ilgooz what do you think of calling `tx.Discard` on error of `tx.Commit`?
The doc of `Commit` says:
> Commit commits the transaction. **If error is not nil, then the transaction is not committed, it can then either be retried or discarded**. Other methods should not be called after transaction has been committed.

Reply:
https://github.com/syndtr/goleveldb/blob/master/leveldb/db_write.go#L263

Look here, they just commit a transaction, and in case of fali discard/retried is not needed

Reply:
> OpenTransaction opens an atomic DB transaction. Only one transaction can be opened at a time. Subsequent call to Write and OpenTransaction will be blocked until in-flight transaction is **committed or discarded**. The returned transaction handle is safe for concurrent use.

👍